File: crm-app/src/workers/eventProcessor.mjs

import { Worker } from 'bullmq';
import IORedis from 'ioredis';
import { logAudit } from '../lib/auditLogger.js';
import { syncChat, applyFacebookLabel, sendFacebookMessage, logAiActionToTimeline } from '../lib/chatService.js';
import { createTask, getCustomerById, getAllEmployees } from '../lib/db.js';
import BusinessAnalyst from '../utils/BusinessAnalyst.js';
import { getOrCreatePersona } from '../lib/personaService.js';
import { sendLineNotify } from '../lib/lineService.js';
import { sendToGoogleSheet } from '../lib/googleSheetsService.js';
import { verifySlip } from '../lib/slipService.js';

const REDIS_URL = process.env.REDIS_URL || 'redis://localhost:6379';
const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
const OWNER_NAME = "คุณบอส (Owner)"; // 👈 เปลี่ยนเป็นชื่อคุณจริงได้ที่นี่
const OWNER_IMAGE = "https://raw.githubusercontent.com/google/material-design-icons/master/png/social/person/materialicons/48dp/1x/baseline_person_black_48dp.png"; // 👈 ใส่ URL รูปโปรไฟล์คุณจริง

const connection = new IORedis(REDIS_URL, {
    maxRetriesPerRequest: null,
});
connection.on('error', (err) => {
    console.warn(`[Redis Worker] Connection error: ${err.message}`);
});

const analyst = GEMINI_API_KEY ? new BusinessAnalyst(GEMINI_API_KEY) : null;

/**
 * Event Processor Worker
 * Consumes jobs from 'fb-events' queue.
 */
const worker = new Worker('fb-events', async (job) => {
    const event = job.data;
    const senderId = event.sender?.id;
    // Map PSID to Conversation ID (Simple assumption for demo)
    const conversationId = `t_${senderId}`;

    const traceId = `TRACE-${job.id}`;

    logAudit({
        action: 'JOB_STARTED',
        actor: 'EventWorker',
        target: senderId || 'Unknown',
        status: 'PENDING',
        traceId,
        details: { jobId: job.id, eventType: 'message' }
    });

    if (!senderId) {
        throw new Error('Invalid Event: Missing Sender ID');
    }

    try {
        // 1. Reactive Sync: Pull latest chat history
        // Sanitize: strip 't_' prefix if present, we need the raw thread ID
        const sanitizedConvId = conversationId.startsWith('t_') ? conversationId.substring(2) : conversationId;

        console.log(`[Worker] Syncing chat for ${sanitizedConvId}...`);
        const chatResult = await syncChat(sanitizedConvId);

        if (chatResult.success) {
            console.log(`[Worker] Chat synced for ${sanitizedConvId}. Source: ${chatResult.source} | Count: ${chatResult.data?.length || 0}`);

            // 🔥 Publish Real-time Update to Redis
            await connection.publish('chat-updates', JSON.stringify({
                conversationId: sanitizedConvId,
                timestamp: new Date().toISOString()
            }));

            // 🤖 AI Proactive Logic (Task + Labels + Persona Reply)
            if (analyst && chatResult.data && chatResult.data.length > 0) {
                const customer = await getCustomerById(sanitizedConvId);
                const customerContext = {
                    name: customer?.profile?.first_name || 'Facebook User',
                    lifecycle_stage: customer?.profile?.lifecycle_stage || 'Lead'
                };

                // A. Task Generation
                console.log(`[Worker] AI analyzing context for proactive tasks...`);
                const suggestedTask = await analyst.suggestProactiveTasks(chatResult.data, customerContext);

                if (suggestedTask) {
                    console.log(`[Worker] AI Suggested Task: ${suggestedTask.title} | Priority: ${suggestedTask.priority}`);
                    const newTask = await createTask({
                        customerId: customer?.customerId || sanitizedConvId,
                        title: `[AI] ${suggestedTask.title}`,
                        description: `${suggestedTask.description}\n\n---\n💡 AI Justification: ${suggestedTask.justification}`,
                        type: suggestedTask.type,
                        priority: suggestedTask.priority,
                        aiGenerated: true,
                        aiContext: {
                            justification: suggestedTask.justification,
                            traceId,
                            actor: 'V-Insight (AI)'
                        }
                    });

                    // Notify UI about new task
                    if (newTask) {
                        await connection.publish('task-updates', JSON.stringify(newTask));
                    } else {
                        console.warn('[Worker] Task creation returned null, skipping Redis publish');
                    }

                    // 🔔 Notify LINE if Priority is HIGH
                    if (newTask && newTask.priority === 'HIGH') {
                        const lineMsg = `\n🚨 [งานด่วน!] พบลูกค้าสนใจสูง\n👤 คุณ: ${customerContext.name}\n📌 เรื่อง: ${newTask.title}\n💡 AI แนะนำ: ${suggestedTask.justification}\n🔗 ดูข้อมูล: ${process.env.NEXT_PUBLIC_BASE_URL || 'http://localhost:3000'}`;
                        await sendLineNotify(lineMsg);
                    }
                }

                // B. Label Generation & Push
                console.log(`[Worker] AI suggesting labels for Meta...`);
                const suggestedLabels = await analyst.suggestLabels(chatResult.data);
                if (suggestedLabels && suggestedLabels.length > 0) {
                    console.log(`[Worker] AI Suggested Labels: ${suggestedLabels.join(', ')}`);
                    for (const label of suggestedLabels) {
                        await applyFacebookLabel(sanitizedConvId, label);
                    }
                }

                // C. Persona Smart Reply (Drafting & Sending)
                // Logic: Only reply if the last message is from the customer and has high intent
                const lastMsg = chatResult.data[chatResult.data.length - 1];
                const isFromCustomer = lastMsg.from?.id === senderId;

                if (isFromCustomer && suggestedTask && suggestedTask.priority === 'HIGH') {
                    console.log(`[Worker] High Intent detected. Drafting Persona reply as ${OWNER_NAME}...`);
                    const replyText = await analyst.generateSmartReply(chatResult.data, OWNER_NAME, customerContext);

                    if (replyText) {
                        const personaId = await getOrCreatePersona(OWNER_NAME, OWNER_IMAGE);
                        const sendResult = await sendFacebookMessage(senderId, replyText, personaId);

                        if (sendResult.success) {
                            console.log(`[Worker] Persona reply sent successfully via ${personaId}`);
                            await logAiActionToTimeline(sanitizedConvId, `Auto-replied as "${OWNER_NAME}" via Persona API`);
                        }
                    }
                }

                // 📊 Export Lead Update to Google Sheet
                await sendToGoogleSheet({
                    type: 'LEAD_UPDATE',
                    customer_id: sanitizedConvId,
                    name: customerContext.name,
                    stage: customerContext.lifecycle_stage,
                    last_message: lastMsg?.message,
                    suggested_labels: suggestedLabels?.join(', '),
                    suggested_task: suggestedTask?.title
                });
            }
        } else {
            console.warn(`[Worker] Chat sync warning: ${chatResult.error}`);
        }

        // 2. Slip Detection
        if (event.attachments && event.attachments[0]?.type === 'image') {
            const imageUrl = event.attachments[0].payload.url;

            logAudit({
                action: 'SLIP_DETECTED',
                actor: 'EventWorker',
                target: senderId,
                status: 'PROCESSING',
                traceId,
                details: { imageUrl }
            });

            const verificationResult = await verifySlip(imageUrl);

            if (verificationResult.verified) {
                logAudit({
                    action: 'SLIP_VERIFIED',
                    actor: 'EventWorker',
                    target: senderId,
                    status: 'SUCCESS',
                    traceId,
                    details: verificationResult.data
                });
                console.log('[Worker] Slip Verified! (Simulated Database Update)');

                // 🔥 Publish Slip Update to Redis
                await connection.publish('slip-updates', JSON.stringify({
                    senderId,
                    verified: true,
                    data: verificationResult.data
                }));

                // 📊 Export Order/Slip to Google Sheet
                await sendToGoogleSheet({
                    type: 'PAYMENT_VERIFIED',
                    customer_id: senderId,
                    amount: verificationResult.data?.amount,
                    bank: verificationResult.data?.receiver?.bank,
                    date: verificationResult.data?.date,
                    status: 'SUCCESS'
                });
            } else {
                logAudit({
                    action: 'SLIP_REJECTED',
                    actor: 'EventWorker',
                    target: senderId,
                    status: 'FAILED',
                    traceId,
                    details: { error: verificationResult.error }
                });
            }
        }

        logAudit({
            action: 'JOB_COMPLETED',
            actor: 'EventWorker',
            target: senderId,
            status: 'SUCCESS',
            traceId
        });

        return { status: 'processed', sender: senderId };

    } catch (error) {
        logAudit({
            action: 'JOB_FAILED',
            actor: 'EventWorker',
            target: senderId,
            status: 'FAILED',
            traceId,
            details: { error: error.message }
        });
        throw error;
    }

}, { connection });

worker.on('completed', (job) => {
    console.log(`[Worker] Job ${job.id} finished!`);
});

worker.on('failed', (job, err) => {
    console.error(`[Worker] Job ${job.id} failed with ${err.message}`);
});

console.log('[Worker] Event Processor started. Listening for jobs...');

File: crm-app/src/workers/python/event_processor.py

import os
import redis
import json
import time
from dotenv import load_dotenv
from PIL import Image
from io import BytesIO
import requests
from google import genai

# Load environment variables
load_dotenv()

# Configuration
REDIS_URL = os.getenv('REDIS_URL', 'redis://localhost:6379')
GEMINI_API_KEY = os.getenv('GEMINI_API_KEY')

# genai configuration removed, now handled via client instantiation

def connect_redis():
    try:
        r = redis.from_url(REDIS_URL)
        r.ping()
        print("[Python Worker] Connected to Redis successfully.")
        return r
    except Exception as e:
        print(f"[Python Worker] Redis connection skipped/failed: {e}")
        return None

from auto_reply import generate_auto_reply

from notification_service import send_staff_notification
from db_adapter import update_customer_intelligence, save_chat_messages, create_task, create_order, add_timeline_event
from behavioral_analyzer import analyze_customer_behavior

def process_event(event):
    """
    Core business logic for processing a Facebook Event.
    """
    sender_id = event.get('sender', {}).get('id')
    if not sender_id:
        return {"success": False, "error": "No sender ID"}

    print(f"[Python Worker] Processing event from: {sender_id}")

    # Initialize results
    intelligence_results = {}

    # 1. Reactive Sync
    print(f"[Python Worker] Syncing chat for {sender_id}...")
    sync_result = sync_chat(f"t_{sender_id}")
    messages = sync_result.get('data', []) if sync_result.get('success') else []
    
    # PHASE 17 & 18: HYBRID TOKEN GUARD (High Intent Real-time + Hourly Batch)
    msg_count = len(messages)
    
    # ⚡ HIGH INTENT (Force Real-time)
    BUY_INTENT_KEYWORDS = ["ราคา", "โอน", "เลขพัสดุ", "สมัคร", "บัญชี", "กี่บาท", "promotion", "โปร", "ส่วนลด", "สน", "จอง"]
    latest_text = event.get('message', '').lower()
    has_buy_intent = any(k in latest_text for k in BUY_INTENT_KEYWORDS)
    
    # Logic: 
    # - Run Intel (Lead Score) every 5 messages or if Buy Intent.
    # - Run Behavioral (Tags) ONLY if Buy Intent (Otherwise wait for Hourly Batch Auditor).
    
    should_run_intel = msg_count <= 1 or msg_count % 5 == 0 or has_buy_intent
    should_run_behavioral = has_buy_intent # Real-time only for sales. Deep profile is handled hourly.

    # AI INTELLIGENCE: Analyze Chat Context (Lead Score/Intent)
    intel_data = None
    if should_run_intel:
        print(f"[Python AI] Running Intel Analysis for {sender_id}...")
        intel_data = analyze_chat_intelligence(sender_id, messages)
        if intel_data:
            intelligence_results['chat'] = intel_data
    else:
        print(f"[Python AI] 🛡️ Skipping Intel Analysis (Token Guard) for {sender_id}")

    # PHASE 16: BEHAVIORAL AI (Deep Tagging & CTA)
    if should_run_behavioral:
        print(f"[Python Worker] Running Behavioral Analysis for {sender_id}...")
        behavioral_data = analyze_customer_behavior(sender_id, messages)
        if behavioral_data and "error" not in behavioral_data:
            intelligence_results['behavioral'] = behavioral_data
            # Update metadata/tags in DB
            update_customer_intelligence(sender_id, {
                "behavioral": behavioral_data,
                "status": behavioral_data.get('customer_status', 'WARM'),
                "tags": behavioral_data.get('behavioral_tags', [])
            })
    else:
        print(f"[Python Worker] 🛡️ Skipping Behavioral Analysis (Token Guard) for {sender_id}")

    # 2. Slip Detection & AI Verification
    attachments = event.get('attachments', [])
    if attachments and attachments[0].get('type') == 'image':
        image_url = attachments[0].get('payload', {}).get('url')
        print(f"[Python Worker] Image detected: {image_url}")
        slip_data = verify_slip_real(sender_id, image_url)
        if slip_data and slip_data.get('status') == 'VERIFIED':
            intelligence_results['slip'] = slip_data
            
            # PHASE 19: Order Persistence
            amount = slip_data.get('amount', 0)
            txn_id = slip_data.get('ref_id') or f"SLIP-{int(time.time())}"
            
            print(f"[Python Worker] 💰 Creating Actual Order for {sender_id}: ฿{amount}")
            create_order(sender_id, txn_id, amount, status="PAID", metadata={"source": "Facebook Slip Detection"})
            add_timeline_event(sender_id, "PURCHASE", f"โอนเงินสำเร็จ ฿{amount}", details=slip_data)

            # Update intel_data for auto-reply logic
            if intel_data: 
                intel_data.update({"intent": "Purchase", "score": 100})
            else:
                intel_data = {"intent": "Purchase", "score": 100, "main_interest": "Payment"}

    # 3. AUTO-REPLY LOGIC (Phase 7)
    auto_reply_text = generate_auto_reply(sender_id, messages, intel_data or {})
    if auto_reply_text:
        send_result = send_facebook_message(sender_id, auto_reply_text)
        intelligence_results['auto_reply'] = {
            "sent": send_result.get('success', False),
            "text": auto_reply_text
        }

    # 4. STAFF ALERTS & TASKS (Phase 7)
    if intel_data:
        score = intel_data.get('score', 0)
        intent = intel_data.get('intent', 'Question')
        
        # High Score Alert (>70)
        if score > 70:
            send_staff_notification(
                title=f"🔥 High Value Lead: {sender_id}",
                message=f"A customer is very interested ({score}%). Intent: {intent}",
                priority="HIGH" if score > 85 else "NORMAL",
                metadata={"ID": sender_id, "Score": score, "Intent": intent}
            )
        
        # Automated Task Generation (>80 or Purchase)
        if score > 80 or intent == 'Purchase':
            task_title = f"Follow-up with {sender_id}"
            task_desc = f"AI assigned high priority. Intent: {intent}, Score: {score}. Please close the sale."
            create_task(sender_id, task_title, task_desc, priority="HIGH" if score > 90 else "NORMAL")
            intelligence_results['automation'] = {"task_created": True}

    return {"success": True, "sender_id": sender_id, "intelligence": intelligence_results}

def send_facebook_message(recipient_id, message_text):
    """
    Send a text message back to the user via Facebook Send API.
    """
    page_access_token = os.getenv('FB_PAGE_ACCESS_TOKEN')
    if not page_access_token:
        return {"success": False, "error": "Missing Page Access Token"}

    url = f"https://graph.facebook.com/v19.0/me/messages?access_token={page_access_token}"
    payload = {
        "recipient": {"id": recipient_id},
        "message": {"text": message_text},
        "messaging_type": "RESPONSE"
    }

    try:
        response = requests.post(url, json=payload)
        result = response.json()
        if response.status_code == 200:
            print(f"[Facebook API] ✅ Message sent to {recipient_id}")
            return {"success": True, "message_id": result.get('message_id')}
        else:
            print(f"[Facebook API] ❌ Send failed: {result.get('error', {}).get('message')}")
            return {"success": False, "error": result.get('error', {}).get('message')}
    except Exception as e:
        print(f"[Facebook API] Error: {e}")
        return {"success": False, "error": str(e)}

from db_adapter import update_customer_intelligence, save_chat_messages

def analyze_chat_intelligence(sender_id, messages):
    """
    Use AI to detect Intent and Lead Score.
    """
    if not GEMINI_API_KEY:
        return None

    try:
        client = genai.Client(api_key=GEMINI_API_KEY)
        
        # Prepare context (last 5 messages)
        chat_text = "\n".join([f"{m.get('from', {}).get('name', 'User')}: {m.get('message', '')}" for m in messages[:5]])
        
        prompt = f"""
        Analyze this conversation for a Japanese Cooking School (V School).
        Conversation:
        {chat_text}
        
        Output only a JSON object with:
        - intent: (one of: Purchase, Question, Complaint, Greeting)
        - score: (integer 0-100 based on conversion probability)
        - main_interest: (short string like 'Sushi Course', 'Price Inquiry', etc.)
        """
        
        response = client.models.generate_content(
            model="gemini-2.0-flash",
            contents=prompt
        )
        # Clean response text
        clean_json = response.text.replace('```json', '').replace('```', '').strip()
        intelligence = json.loads(clean_json)
        
        print(f"[Python AI] Lead Score for {sender_id}: {intelligence.get('score')}")
        
        # Use DB Adapter instead of direct fs
        update_customer_intelligence(sender_id, intelligence)
        return intelligence
        
    except Exception as e:
        print(f"[Python AI] Intelligence Error: {e}")
        return None

def verify_slip_real(sender_id, image_url):
    """
    Real OCR/AI Verification using Gemini Vision.
    """
    if not GEMINI_API_KEY:
        return None

    try:
        client = genai.Client(api_key=GEMINI_API_KEY)

        # Download image
        response = requests.get(image_url)
        img = Image.open(BytesIO(response.content))

        prompt = """
        Analyze this Thai bank transfer slip. Extract the following information in JSON format:
        - bank_name
        - amount (float)
        - date (YYYY-MM-DD)
        - time (HH:mm)
        - ref_id (Transaction ID)
        - is_valid (Boolean - check for standard Bank Layout)
        """

        response = client.models.generate_content(
            model="gemini-2.0-flash", # gemini-2.0-flash handles multimodal
            contents=[prompt, img]
        )
        clean_json = response.text.replace('```json', '').replace('```', '').strip()
        result = json.loads(clean_json)
        
        print(f"[Python AI] ✅ Slip Analyzed: {result.get('amount')} THB via {result.get('bank_name')}")
        
        result['status'] = 'VERIFIED' if result.get('is_valid') else 'REJECTED'
        
        # Use DB Adapter instead of direct fs
        intel_update = {
            "slip_verification": result, 
            "score": 100 if result.get('is_valid') else 0, 
            "intent": "Purchase"
        }
        update_customer_intelligence(sender_id, intel_update)
        return result

    except Exception as e:
        print(f"[Python AI] Verification Failed: {e}")
        return None

def sync_chat(conversation_id):
    """
    Fetch messages from Facebook and save to local cache or DB.
    """
    page_access_token = os.getenv('FB_PAGE_ACCESS_TOKEN')
    if not page_access_token:
        return {'success': False, 'error': 'Missing Page Access Token'}

    url = f"https://graph.facebook.com/v19.0/{conversation_id}/messages"
    params = {
        'fields': 'id,message,from,created_time,attachments{id,mime_type,name,file_url,image_data,url}',
        'limit': 50,
        'access_token': page_access_token
    }

    try:
        response = requests.get(url, params=params)
        data = response.json()

        if response.status_code == 200:
            messages = data.get('data', [])
            # Use DB Adapter
            save_chat_messages(conversation_id, messages)
            return {'success': True, 'data': messages, 'source': 'facebook'}
        else:
            return {'success': False, 'error': data.get('error', {}).get('message')}
    except Exception as e:
        return {'success': False, 'error': str(e)}

def main():
    direct_input = os.getenv('PYTHON_INPUT')
    if direct_input:
        try:
            event = json.loads(direct_input)
            result = process_event(event)
            print(json.dumps(result))
        except Exception as e:
            print(json.dumps({"success": False, "error": str(e)}))
        return

    print("[Python Worker] Starting in Queue Mode...")
    r = connect_redis()
    while True:
        # Queue logic would go here
        time.sleep(1)

if __name__ == "__main__":
    main()
